lca-chimevc-stack/lambda_functions/call_transcriber/index.js

```diff
@@ -535,206 +693,137 @@ const go = async function (
-    }
-    else {
-      //agentStreamArn = event.detail.streamArn;
-      console.log("this is not the caller stream, so return.");
+  console.log("Event: ", JSON.stringify(event));
+
+  // initialise clients (globals) each invocation to avoid possibility of ECONNRESET in subsequent invocations.
+  s3Client = new S3Client({ region: REGION });
+  lambdaClient = new LambdaClient({ region: REGION });
+  dynamoClient = new DynamoDBClient({ region: REGION });
+  kinesisClient = new KinesisClient({ region: REGION });
+
+  /*
+  Create a callData object for incoming event:
+  A LAMBDA_CONTINUE event contains callData object ready to use
+  A START_CALL_PROCESSING event contains callId which is used to look up a previously stored callData object
+  Chime stream STARTED events for both AGENT and CALLER streams are combined to create a new callData object
+    - the AGENT event is stored to DynamoDB and the function exits
+    - the CALLER event is correlated with stored AGENT event to create callData object
+    - an optional user defined Lambda hook may:
+         - manipulate callData object fields
+         - save callData object to DynamoDB and delay or disable call processing until/if a START_CALL_PROCESSING is received later
+  */
+
+  let callData;
+
+  if (event['action'] === 'LAMBDA_CONTINUE') {
+    callData = event.callData;
+    console.log('--- CONTINUING FROM PREVIOUS LAMBDA. LAMBDA SEQUENCE COUNT: ', callData.lambdaCount, '---');
+    if (callData.lambdaCount > 30) {
+      console.log('Stopping due to runaway recursive Lambda.');
       return;
     }
+  }
 
-    let streamResults = await getStreamsFromDynamo(
-      event.detail.callId,
-      agentStreamArn,
-      callerStreamArn,
-    );
-    console.log(`agent stream:${streamResults.agentStreamArn}`);
-    console.log(`caller stream:${streamResults.callerStreamArn}`);
-
-    let loopCount = 0;
-
-    while (streamResults.agentStreamArn === undefined || streamResults.callerStreamArn === undefined) {
-      console.log(loopCount,'Agent or caller streams not yet available. Sleeping 100ms.');
-      await sleep(100);
-      streamResults = await getStreamsFromDynamo(
-        event.detail.callId,
-        agentStreamArn,
-        callerStreamArn,
-      );
-      console.log(`agent stream:${streamResults.agentStreamArn}`);
-      console.log(`caller stream:${streamResults.callerStreamArn}`);
-      loopCount = loopCount + 1;
-      if(loopCount == 100) {
-        console.log("Both KVS streams not active after 10 seconds. Exiting.");
-        return;
-      }
+  else if (event['source'] === 'lca-solution' && event['detail-type'] === 'START_CALL_PROCESSING') {
+    console.log('START_CALL_PROCESSING event received, Retrieving previously stored callData.');
+    callData = await getCallDataForStartCallProcessingEvent(event.detail);
+    if (!callData) {
+      console.log('Nothing to do - exiting.');
+      return;
     }
+    await writeCallDataToDdb(callData);
```

Comment:
This writing call to dynamo overwrites what it just wrote in the getCallDataForStartCallProcessingEvent, except it contains updated info from the scpEvent. Rather than write twice, why not just write once at the end of the getCallDataForStartCallProcessingEvent?  not a blocker, though.

Reply:
Yes, on inspection you're right. Will remove the call to writeCallDataToDdb() that's inside getCallDataForStartCallProcessingEvent(), and just let it get stored from the highlighted line (in the main handler) instead - consistent with all other calls to writeCallDataToDdb(). Tx.